libursa/Database.cpp

```diff
@@ -239,23 +262,30 @@ void Database::commit_task(uint64_t task_id) {
         }
     }
 
-    if (!task->changes.empty()) {
+    if (!changes.empty()) {
         save();
     }
 
     erase_task(task_id);
 }
 
-Task *Database::get_task(uint64_t task_id) { return tasks.at(task_id).get(); }
+const TaskSpec &Database::get_task(uint64_t task_id) {
+    return *tasks.at(task_id).get();
+}
 
 void Database::erase_task(uint64_t task_id) { tasks.erase(task_id); }
 
 DatabaseSnapshot Database::snapshot() {
     std::vector<const OnDiskDataset *> cds;
-
     for (const auto *d : working_datasets) {
         cds.push_back(d);
     }
 
-    return DatabaseSnapshot(db_name, db_base, config, iterators, cds, tasks);
+    std::unordered_map<uint64_t, TaskSpec *> taskspecs;
+    for (const auto &[k, v] : tasks) {
+        taskspecs.emplace(k, v.get());
+    }
+
+    return DatabaseSnapshot(db_name, db_base, config, iterators, cds,
+                            taskspecs);
```

Comment:
Maybe it would be better to push this into DatabaseSnapshot class and add a constructor `DatabaseSnapshot::DatabaseSnaphot(const Database &)`

Reply:
I'm trying to keep public interfaces small, and this would mean that database has to expose accessors for everything - so maybe keep it as it is?

Reply:
not really, we can declare DatabaseSnapshot as friend. What do you think?